demos/common/hfsm/az_iot_hfsm.c

```diff
@@ -0,0 +1,324 @@
+      {
+        this_iothfsm->_retry_attempt++;
+      }
+
+      az_iot_hfsm_event_data_error* error_data = (az_iot_hfsm_event_data_error*)(event.data);
+
+      bool should_retry = false;
+      if (error_data->type == AZ_IOT_ERROR_TYPE_SERVICE)
+      {
+        should_retry = az_iot_status_retriable(error_data->iot_status);
+      }
+      else if (error_data->type == AZ_IOT_ERROR_TYPE_NETWORK)
+      {
+        should_retry = true;
+      }
+
+      if (should_retry)
+      {
+        int32_t random_jitter_msec = az_iot_hfsm_pal_get_random_jitter_msec(me);
+
+        int32_t retry_delay_msec = az_iot_calculate_retry_delay(
+          operation_msec, 
+          this_iothfsm->_retry_attempt,
+          AZ_IOT_HFSM_MIN_RETRY_DELAY_MSEC,
+          AZ_IOT_HFSM_MAX_RETRY_DELAY_MSEC,
+          random_jitter_msec);
+
+        LogWarn( ("AzureIoT: AZ_IOT_ERROR retrying after %" PRId32 "ms", retry_delay_msec) );
+        ret = az_hfsm_pal_timer_start(me, this_iothfsm->_timer_handle, retry_delay_msec, true);
+      }
+      else
+      {
+        this_iothfsm->_use_secondary_credentials = !this_iothfsm->_use_secondary_credentials;
+        this_iothfsm->_start_time_msec = az_hfsm_pal_timer_get_milliseconds();
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+          ret = az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start);
+          ret = az_hfsm_transition_substate(me, azure_iot, provisioning);
+#else
+          ret = az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start);
+          ret = az_hfsm_transition_substate(me, azure_iot, hub);
+#endif
+      }
+      break;
+
+    case AZ_HFSM_EXIT:
+    case AZ_HFSM_ERROR:
+    case AZ_HFSM_TIMEOUT:
+    default:
+      LogInfo( ("AzureIoT: PANIC!") );
+      az_iot_hfsm_pal_critical_error(me);
+      break;
+  }
+
+  return ret;
+}
+
+static int32_t idle(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = 0;
+  az_iot_hfsm_type* this_iothfsm = (az_iot_hfsm_type*)me;
+
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+      LogInfo( ("idle: Entry") );
+      break;
+
+    case AZ_HFSM_EXIT:
+      LogInfo( ("idle: Exit") );
+      break;
+    
+    case AZ_IOT_START:
+      LogInfo( ("idle: AZ_IOT_START") );
+
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+      if(az_hfsm_post_event(this_iothfsm->_provisioning_hfsm, az_hfsm_event_az_iot_start))
+      {
+        LogError( ("idle: az_hfsm_post_event to _provisioning_hfsm failed.") );
+        ret = az_hfsm_post_event(me, az_hfsm_errork_unknown_event);
+      }
+      else
+      {
+        ret = az_hfsm_transition_peer(me, idle, provisioning);
+      }
+#else
+      if(az_hfsm_post_event(this_iothfsm->_iothub_hfsm, az_hfsm_event_az_iot_start))
+      {
+        LogError( ("idle: az_hfsm_post_event to _iothub_hfsm failed.") );
+        ret = az_hfsm_post_event(me, az_hfsm_errork_unknown_event);
+      }
+      else
+      {
+        ret = az_hfsm_transition_peer(me, idle, hub);
+      }
+#endif
```

Comment:
pardon my ignorance here. but if I have `AZ_IOT_HFSM_PROVISIONING_ENABLED`, is there never a case where I would make the transition out of provisioning and into hub and need to idle in `this_iothfsm->_iothub_hfsm`? Should 200-208 really be stripped out? I'm imagining after line 248 executes

Reply:
The two diagrams are:
1. With provisioning:
 
![image](https://user-images.githubusercontent.com/8635911/124035788-e1765700-d9b1-11eb-8f1c-24f5e31e197d.png)

2. Without provisioning:
![image](https://user-images.githubusercontent.com/8635911/124035847-f6eb8100-d9b1-11eb-868c-d6ddcba5b38a.png)

So in the first case, the action is to post an event to the actual provisioning engine to start then transition into the "provisioning" state. In the second, the action is to send the hub engine the request to start then transition into the "hub" state.

`ret = az_hfsm_transition_peer(me, idle, hub);` means transition "me" (HFSM) from the "idle" source state to the "hub" destination state.

The current state for _iothub_hfsm is private and is managed only by the code in that respective HFSM. Indeed, it's expected that HFSM is going to be inactive (idling) in an error condition until it receives an event that transitions it into another state.